File: sdks/python/reproit_sdk/managed_closure.py
"""The static managed capture closure: world binding and frozen artifacts.

Mirrors the closure half of crates/reproit-sdk-rust/src/managed.rs: the
world checkpoint shape the SDK consumes, the static artifact set proof,
dependency-transcript validation, and freezing artifact bytes into a
private spool so they cannot change between proof and upload.
"""

from __future__ import annotations

import hashlib
import os
import stat
import tempfile
from collections.abc import Mapping
from dataclasses import dataclass

from reproit_sdk import canonical_bytes

from .managed_protocol import (
    DEPENDENCY_TRANSCRIPT_MEDIA_TYPE,
    MAX_CHUNK_BYTES,
    ManagedError,
    canonical_digest,
    incomplete_candidate,
    new_object_id,
    parse_strict_json,
    schema_invalid,
    valid_digest,
    valid_timestamp,
    valid_typed_id,
)

MAX_CAPTURE_ARTIFACT_BYTES = 274_878_824_448
MAX_WORLD_MANIFEST_BYTES = 262_144
COPY_BUFFER_BYTES = 64 * 1024

ARTIFACT_ROLES = frozenset(("dependency-transcript", "world-state"))


@dataclass(frozen=True)
class ManagedCandidateArtifact:
    media_type: str
    object_id: str
    path: str
    role: str
    uri: str


@dataclass
class ManagedCaptureClosure:
    """The static capture closure the application proves before upload."""

    artifacts: list[ManagedCandidateArtifact]
    completion: str
    world: dict[str, object]


class FrozenManagedCaptureClosure:
    """A capture closure whose artifact bytes are frozen in a private spool."""

    def __init__(self, closure: ManagedCaptureClosure):
        validate_world_checkpoint(closure.world)
        validate_static_artifact_set(closure.world, closure.artifacts)
        self._spool: tempfile.TemporaryDirectory | None = None
        artifacts = closure.artifacts
        if artifacts:
            self._spool = tempfile.TemporaryDirectory(prefix="reproit-managed-world-")
            artifacts = [
                _freeze_artifact(artifact, self._spool.name)
                for artifact in closure.artifacts
            ]
        validate_static_artifact_set(closure.world, artifacts)
        self.closure = ManagedCaptureClosure(
            artifacts, closure.completion, closure.world
        )

    def world_id(self) -> str:
        validate_world_checkpoint(self.closure.world)
        return canonical_digest(self.closure.world)


def validate_world_checkpoint(value: object) -> None:
    """Validate the bounded world checkpoint shape the SDK consumes."""
    if (
        not isinstance(value, Mapping)
        or set(value) != {"created_at", "format", "points"}
        or value["format"] != "reproit.world-checkpoint.v1"
        or not valid_timestamp(value["created_at"])
        or not isinstance(value["points"], list)
        or len(value["points"]) > 64
    ):
        raise schema_invalid()
    providers = set()
    for point in value["points"]:
        if (
            not isinstance(point, Mapping)
            or point.get("format") != "reproit.recoverable-point.v1"
            or not isinstance(point.get("provider_id"), str)
            or not isinstance(point.get("artifacts"), list)
            or len(point["artifacts"]) > 32_767
            or point["provider_id"] in providers
        ):
            raise schema_invalid()
        providers.add(point["provider_id"])
        for artifact in point["artifacts"]:
            if (
                not isinstance(artifact, Mapping)
                or not valid_digest(artifact.get("digest"))
                or type(artifact.get("size")) is not int
                or artifact["size"] < 0
                or not isinstance(artifact.get("uri"), str)
                or not artifact["uri"]
                or len(artifact["uri"]) > 2_048
                or not isinstance(artifact.get("media_type"), str)
            ):
                raise schema_invalid()
    if len(canonical_bytes(value)) > MAX_WORLD_MANIFEST_BYTES:
        raise schema_invalid()


def expected_world_artifacts(
    world: Mapping[str, object],
) -> set[tuple[str, str, int, str]]:
    return {
        (artifact["uri"], artifact["digest"], artifact["size"], artifact["media_type"])
        for point in world["points"]
        for artifact in point["artifacts"]
    }


def validate_static_artifact_set(
    world: Mapping[str, object], artifacts: list[ManagedCandidateArtifact]
) -> None:
    if len(artifacts) > 32_767:
        raise incomplete_candidate()
    expected_world = expected_world_artifacts(world)
    supplied_world = {
        artifact.uri for artifact in artifacts if artifact.role == "world-state"
    }
    if len(expected_world) != len(supplied_world) or any(
        uri not in supplied_world for uri, _, _, _ in expected_world
    ):
        raise incomplete_candidate()
    object_ids: set[str] = set()
    uris: set[str] = set()
    for artifact in artifacts:
        if (
            artifact.role not in ARTIFACT_ROLES
            or not artifact.uri
            or len(artifact.uri) > 2_048
            or not artifact.media_type
            or len(artifact.media_type) > 256
            or artifact.object_id in object_ids
            or artifact.uri in uris
        ):
            raise incomplete_candidate()
        object_ids.add(artifact.object_id)
        uris.add(artifact.uri)
        size, digest = hash_file(artifact.path)
        if (
            artifact.role == "world-state"
            and (
                artifact.uri,
                digest,
                size,
                artifact.media_type,
            )
            not in expected_world
        ):
            raise incomplete_candidate()
        if (
            artifact.role == "dependency-transcript"
            and artifact.media_type == DEPENDENCY_TRANSCRIPT_MEDIA_TYPE
        ):
            if size > MAX_CHUNK_BYTES:
                raise incomplete_candidate()
            validate_transcript_bytes(read_bounded(artifact.path, size))


def validate_transcript_bytes(value: bytes) -> dict[str, object]:
    """Mirror the DependencyTranscript strict parse and validation."""
    parsed = parse_strict_json(value, MAX_CHUNK_BYTES)
    if (
        not isinstance(parsed, dict)
        or canonical_bytes(parsed) != value
        or set(parsed) != {"adapter_id", "adapter_version", "format", "interactions"}
        or parsed["format"] != "reproit.dependency-transcript.v1"
    ):
        raise schema_invalid()
    adapter_id = parsed["adapter_id"]
    adapter_version = parsed["adapter_version"]
    interactions = parsed["interactions"]
    if (
        not isinstance(adapter_id, str)
        or not adapter_id
        or len(adapter_id) > 128
        or not isinstance(adapter_version, str)
        or not adapter_version
        or len(adapter_version) > 64
        or not isinstance(interactions, list)
        or not 1 <= len(interactions) <= 1_024
    ):
        raise schema_invalid()
    for index, interaction in enumerate(interactions):
        _validate_interaction(interaction, index)
    return parsed


def _validate_interaction(interaction: object, index: int) -> None:
    if (
        not isinstance(interaction, Mapping)
        or set(interaction)
        != {
            "causal_parent_id",
            "operation_id",
            "outcome",
            "request_digest",
            "request_object_id",
            "response_digest",
            "response_object_id",
            "sequence",
            "session_position",
        }
        or interaction["sequence"] != index
        or not valid_typed_id(interaction["operation_id"], "operation_id")
        or not (
            interaction["causal_parent_id"] is None
            or valid_typed_id(interaction["causal_parent_id"], "operation_id")
        )
        or not valid_digest(interaction["request_digest"])
        or not valid_digest(interaction["response_digest"])
        or not valid_typed_id(interaction["request_object_id"], "object_id")
        or not valid_typed_id(interaction["response_object_id"], "object_id")
        or type(interaction["session_position"]) is not int
        or not 0 <= interaction["session_position"] <= 9_007_199_254_740_991
    ):
        raise schema_invalid()


def _freeze_artifact(
    artifact: ManagedCandidateArtifact, spool_path: str
) -> ManagedCandidateArtifact:
    metadata = artifact_metadata(artifact.path)
    temporary = os.path.join(spool_path, f"artifact-{new_object_id()}")
    first_digest, copied = _copy_and_digest(artifact.path, temporary, metadata.st_size)
    second_digest, verified = _digest_file(artifact.path, metadata.st_size)
    if first_digest != second_digest or copied != verified:
        raise incomplete_candidate()
    frozen_path = os.path.join(spool_path, first_digest.removeprefix("sha256:"))
    if os.path.exists(frozen_path):
        stored_digest, stored_size = _digest_file(frozen_path, copied)
        if stored_digest != first_digest or stored_size != copied:
            raise ManagedError(
                "OBJECT_DIGEST_MISMATCH",
                "The object bytes do not match the bound digest.",
            )
        os.remove(temporary)
    else:
        os.replace(temporary, frozen_path)
    return ManagedCandidateArtifact(
        artifact.media_type,
        artifact.object_id,
        frozen_path,
        artifact.role,
        artifact.uri,
    )


def artifact_metadata(path: str) -> os.stat_result:
    try:
        metadata = os.lstat(path)
    except OSError as error:
        raise incomplete_candidate() from error
    if (
        not stat.S_ISREG(metadata.st_mode)
        or stat.S_ISLNK(metadata.st_mode)
        or metadata.st_size > MAX_CAPTURE_ARTIFACT_BYTES
    ):
        raise incomplete_candidate()
    return metadata


def _copy_and_digest(source: str, target: str, expected: int) -> tuple[str, int]:
    hasher = hashlib.sha256()
    total = 0
    try:
        with open(source, "rb") as reader, open(target, "wb") as writer:
            while True:
                chunk = reader.read(COPY_BUFFER_BYTES)
                if not chunk:
                    break
                total += len(chunk)
                if total > expected:
                    raise incomplete_candidate()
                hasher.update(chunk)
                writer.write(chunk)
            writer.flush()
    except OSError as error:
        raise ManagedError(
            "SERVICE_UNAVAILABLE",
            "Repro It could not create the bounded local ciphertext staging area.",
        ) from error
    if total != expected:
        raise incomplete_candidate()
    return f"sha256:{hasher.hexdigest()}", total


def _digest_file(path: str, expected: int) -> tuple[str, int]:
    hasher = hashlib.sha256()
    total = 0
    try:
        with open(path, "rb") as reader:
            while True:
                chunk = reader.read(COPY_BUFFER_BYTES)
                if not chunk:
                    break
                total += len(chunk)
                if total > expected:
                    raise incomplete_candidate()
                hasher.update(chunk)
    except OSError as error:
        raise incomplete_candidate() from error
    if total != expected:
        raise incomplete_candidate()
    return f"sha256:{hasher.hexdigest()}", total


def read_bounded(path: str, expected: int) -> bytes:
    try:
        with open(path, "rb") as source:
            value = source.read(expected + 1)
    except OSError as error:
        raise incomplete_candidate() from error
    if len(value) != expected:
        raise incomplete_candidate()
    return value


def hash_file(path: str) -> tuple[int, str]:
    """Hash a stable regular file, failing closed if it changes underneath."""
    before = artifact_metadata(path)
    digest, size = _digest_file(path, before.st_size)
    after = artifact_metadata(path)
    if after.st_size != size or after.st_mtime_ns != before.st_mtime_ns:
        raise incomplete_candidate()
    return size, digest

File: sdks/node/src/managed-closure.js
// The static managed capture closure: world binding and frozen artifacts.
//
// Mirrors the closure half of crates/reproit-sdk-rust/src/managed.rs: the
// world checkpoint shape the SDK consumes, the static artifact set proof,
// dependency-transcript validation, and freezing artifact bytes into a
// private spool so they cannot change between proof and upload.

import { Buffer } from "node:buffer";
import { createHash } from "node:crypto";
import {
  closeSync,
  copyFileSync,
  existsSync,
  lstatSync,
  mkdtempSync,
  openSync,
  readSync,
  renameSync,
  rmSync,
  unlinkSync,
} from "node:fs";
import os from "node:os";
import path from "node:path";
import { releaseLogical, reserveLogical } from "./process-resources.js";

import { canonicalBytes } from "./index.js";
import {
  DEPENDENCY_TRANSCRIPT_MEDIA_TYPE,
  MAX_CHUNK_BYTES,
  ManagedError,
  canonicalDigest,
  incompleteCandidate,
  isObject,
  newObjectId,
  parseStrictJson,
  sameKeys,
  schemaInvalid,
  validDigest,
  validTimestamp,
  validTypedId,
} from "./managed-protocol.js";

export const MAX_CAPTURE_ARTIFACT_BYTES = 1024 * 1024 * 1024;
export const MAX_WORLD_ARTIFACT_BYTES = 2 * 1024 * 1024 * 1024;
export const MAX_WORLD_MANIFEST_BYTES = 262_144;
export const COPY_BUFFER_BYTES = 64 * 1024;

export const ARTIFACT_ROLES = new Set([
  "dependency-transcript",
  "world-state",
]);

// One static capture artifact: { mediaType, objectId, path, role, uri }.
// The plain-object shape mirrors the Python and Rust artifact records.

// A capture closure whose artifact bytes are frozen in a private spool.
export class FrozenManagedCaptureClosure {
  #spool = null;
  #reservedBytes = 0;

  constructor(closure) {
    validateWorldCheckpoint(closure.world);
    validateStaticArtifactSet(closure.world, closure.artifacts);
    this.#reservedBytes = closure.artifacts.reduce((total, artifact) => {
      const size = artifactMetadata(artifact.path).size;
      if (total > MAX_WORLD_ARTIFACT_BYTES - size) {
        throw incompleteCandidate();
      }
      return total + size;
    }, 0);
    if (!reserveLogical(this.#reservedBytes)) throw incompleteCandidate();
    let artifacts = closure.artifacts;
    if (artifacts.length > 0) {
      this.#spool = mkdtempSync(
        path.join(os.tmpdir(), "reproit-managed-world-"),
      );
      try {
        artifacts = closure.artifacts.map((artifact) =>
          freezeArtifact(artifact, this.#spool),
        );
      } catch (error) {
        this.dispose();
        throw error;
      }
    }
    validateStaticArtifactSet(closure.world, artifacts);
    this.closure = {
      artifacts,
      completion: closure.completion,
      world: closure.world,
    };
  }

  worldId() {
    validateWorldCheckpoint(this.closure.world);
    return canonicalDigest(this.closure.world);
  }

  dispose() {
    if (this.#spool !== null) {
      rmSync(this.#spool, { force: true, recursive: true });
      this.#spool = null;
    }
    if (this.#reservedBytes > 0) {
      releaseLogical(this.#reservedBytes);
      this.#reservedBytes = 0;
    }
  }
}

// Validate the bounded world checkpoint shape the SDK consumes.
export function validateWorldCheckpoint(value) {
  if (
    !sameKeys(value, ["created_at", "format", "points"]) ||
    value.format !== "reproit.world-checkpoint.v1" ||
    !validTimestamp(value.created_at) ||
    !Array.isArray(value.points) ||
    value.points.length > 64
  ) {
    throw schemaInvalid();
  }
  const providers = new Set();
  for (const point of value.points) {
    if (
      !isObject(point) ||
      point.format !== "reproit.recoverable-point.v1" ||
      typeof point.provider_id !== "string" ||
      !Array.isArray(point.artifacts) ||
      point.artifacts.length > 32_767 ||
      providers.has(point.provider_id)
    ) {
      throw schemaInvalid();
    }
    providers.add(point.provider_id);
    for (const artifact of point.artifacts) {
      if (
        !isObject(artifact) ||
        !validDigest(artifact.digest) ||
        !Number.isSafeInteger(artifact.size) ||
        artifact.size < 0 ||
        typeof artifact.uri !== "string" ||
        artifact.uri.length === 0 ||
        artifact.uri.length > 2_048 ||
        typeof artifact.media_type !== "string"
      ) {
        throw schemaInvalid();
      }
    }
  }
  if (canonicalBytes(value).length > MAX_WORLD_MANIFEST_BYTES) {
    throw schemaInvalid();
  }
}

export function expectedWorldArtifacts(world) {
  const expected = new Set();
  for (const point of world.points) {
    for (const artifact of point.artifacts) {
      expected.add(
        JSON.stringify([
          artifact.uri,
          artifact.digest,
          artifact.size,
          artifact.media_type,
        ]),
      );
    }
  }
  return expected;
}

export function validateStaticArtifactSet(world, artifacts) {
  if (artifacts.length > 32_767) {
    throw incompleteCandidate();
  }
  const expectedWorld = expectedWorldArtifacts(world);
  const suppliedWorld = new Set(
    artifacts
      .filter((artifact) => artifact.role === "world-state")
      .map((artifact) => artifact.uri),
  );
  const expectedUris = new Set(
    [...expectedWorld].map((entry) => JSON.parse(entry)[0]),
  );
  if (
    expectedUris.size !== suppliedWorld.size ||
    [...expectedUris].some((uri) => !suppliedWorld.has(uri))
  ) {
    throw incompleteCandidate();
  }
  const objectIds = new Set();
  const uris = new Set();
  for (const artifact of artifacts) {
    if (
      !ARTIFACT_ROLES.has(artifact.role) ||
      typeof artifact.uri !== "string" ||
      artifact.uri.length === 0 ||
      artifact.uri.length > 2_048 ||
      typeof artifact.mediaType !== "string" ||
      artifact.mediaType.length === 0 ||
      artifact.mediaType.length > 256 ||
      objectIds.has(artifact.objectId) ||
      uris.has(artifact.uri)
    ) {
      throw incompleteCandidate();
    }
    objectIds.add(artifact.objectId);
    uris.add(artifact.uri);
    const { size, digest } = hashFile(artifact.path);
    if (
      artifact.role === "world-state" &&
      !expectedWorld.has(
        JSON.stringify([artifact.uri, digest, size, artifact.mediaType]),
      )
    ) {
      throw incompleteCandidate();
    }
    if (
      artifact.role === "dependency-transcript" &&
      artifact.mediaType === DEPENDENCY_TRANSCRIPT_MEDIA_TYPE
    ) {
      if (size > MAX_CHUNK_BYTES) {
        throw incompleteCandidate();
      }
      validateTranscriptBytes(readBounded(artifact.path, size));
    }
  }
}

// Mirror the DependencyTranscript strict parse and validation.
export function validateTranscriptBytes(value) {
  const parsed = parseStrictJson(value, MAX_CHUNK_BYTES);
  if (
    !isObject(parsed) ||
    !canonicalBytes(parsed).equals(Buffer.from(value)) ||
    !sameKeys(parsed, [
      "adapter_id",
      "adapter_version",
      "format",
      "interactions",
    ]) ||
    parsed.format !== "reproit.dependency-transcript.v1"
  ) {
    throw schemaInvalid();
  }
  const adapterId = parsed.adapter_id;
  const adapterVersion = parsed.adapter_version;
  const interactions = parsed.interactions;
  if (
    typeof adapterId !== "string" ||
    adapterId.length === 0 ||
    adapterId.length > 128 ||
    typeof adapterVersion !== "string" ||
    adapterVersion.length === 0 ||
    adapterVersion.length > 64 ||
    !Array.isArray(interactions) ||
    interactions.length < 1 ||
    interactions.length > 1_024
  ) {
    throw schemaInvalid();
  }
  interactions.forEach((interaction, index) => {
    validateInteraction(interaction, index);
  });
  return parsed;
}

function validateInteraction(interaction, index) {
  if (
    !sameKeys(interaction, [
      "causal_parent_id",
      "operation_id",
      "outcome",
      "request_digest",
      "request_object_id",
      "response_digest",
      "response_object_id",
      "sequence",
      "session_position",
    ]) ||
    interaction.sequence !== index ||
    !validTypedId(interaction.operation_id, "operation_id") ||
    !(
      interaction.causal_parent_id === null ||
      validTypedId(interaction.causal_parent_id, "operation_id")
    ) ||
    !validDigest(interaction.request_digest) ||
    !validDigest(interaction.response_digest) ||
    !validTypedId(interaction.request_object_id, "object_id") ||
    !validTypedId(interaction.response_object_id, "object_id") ||
    !Number.isSafeInteger(interaction.session_position) ||
    interaction.session_position < 0
  ) {
    throw schemaInvalid();
  }
}

function freezeArtifact(artifact, spoolPath) {
  const metadata = artifactMetadata(artifact.path);
  const temporary = path.join(spoolPath, `artifact-${newObjectId()}`);
  const { digest: firstDigest, size: copied } = copyAndDigest(
    artifact.path,
    temporary,
    metadata.size,
  );
  const { digest: secondDigest, size: verified } = digestFile(
    artifact.path,
    metadata.size,
  );
  if (firstDigest !== secondDigest || copied !== verified) {
    throw incompleteCandidate();
  }
  const frozenPath = path.join(
    spoolPath,
    firstDigest.replace("sha256:", ""),
  );
  if (existsSync(frozenPath)) {
    const { digest: storedDigest, size: storedSize } = digestFile(
      frozenPath,
      copied,
    );
    if (storedDigest !== firstDigest || storedSize !== copied) {
      throw new ManagedError(
        "OBJECT_DIGEST_MISMATCH",
        "The object bytes do not match the bound digest.",
      );
    }
    unlinkSync(temporary);
  } else {
    renameSync(temporary, frozenPath);
  }
  return {
    mediaType: artifact.mediaType,
    objectId: artifact.objectId,
    path: frozenPath,
    role: artifact.role,
    uri: artifact.uri,
  };
}

export function artifactMetadata(filePath) {
  let metadata;
  try {
    metadata = lstatSync(filePath);
  } catch {
    throw incompleteCandidate();
  }
  if (
    !metadata.isFile() ||
    metadata.isSymbolicLink() ||
    metadata.size > MAX_CAPTURE_ARTIFACT_BYTES
  ) {
    throw incompleteCandidate();
  }
  return metadata;
}

function copyAndDigest(source, target, expected) {
  try {
    copyFileSync(source, target);
  } catch {
    throw new ManagedError(
      "SERVICE_UNAVAILABLE",
      "Repro It could not create the bounded local ciphertext staging area.",
    );
  }
  const copy = digestFile(target, expected);
  if (copy.size !== expected) {
    throw incompleteCandidate();
  }
  return copy;
}

export function digestFile(filePath, expected) {
  const hasher = createHash("sha256");
  let total = 0;
  let descriptor;
  try {
    descriptor = openSync(filePath, "r");
  } catch {
    throw incompleteCandidate();
  }
  try {
    const buffer = Buffer.alloc(COPY_BUFFER_BYTES);
    for (;;) {
      const read = readSync(descriptor, buffer, 0, buffer.length, null);
      if (read === 0) break;
      total += read;
      if (total > expected) {
        throw incompleteCandidate();
      }
      hasher.update(buffer.subarray(0, read));
    }
  } catch (error) {
    if (error instanceof ManagedError) throw error;
    throw incompleteCandidate();
  } finally {
    closeSync(descriptor);
  }
  if (total !== expected) {
    throw incompleteCandidate();
  }
  return { digest: `sha256:${hasher.digest("hex")}`, size: total };
}

export function readBounded(filePath, expected) {
  let descriptor;
  try {
    descriptor = openSync(filePath, "r");
  } catch {
    throw incompleteCandidate();
  }
  try {
    const buffer = Buffer.alloc(expected + 1);
    let total = 0;
    for (;;) {
      const read = readSync(
        descriptor,
        buffer,
        total,
        buffer.length - total,
        null,
      );
      if (read === 0) break;
      total += read;
      if (total >= buffer.length) break;
    }
    if (total !== expected) {
      throw incompleteCandidate();
    }
    return Buffer.from(buffer.subarray(0, expected));
  } catch (error) {
    if (error instanceof ManagedError) throw error;
    throw incompleteCandidate();
  } finally {
    closeSync(descriptor);
  }
}

// Hash a stable regular file, failing closed if it changes underneath.
export function hashFile(filePath) {
  const before = artifactMetadata(filePath);
  const beforeStat = lstatSync(filePath, { bigint: true });
  const { digest, size } = digestFile(filePath, before.size);
  const after = artifactMetadata(filePath);
  const afterStat = lstatSync(filePath, { bigint: true });
  if (after.size !== size || afterStat.mtimeNs !== beforeStat.mtimeNs) {
    throw incompleteCandidate();
  }
  return { digest, size };
}
